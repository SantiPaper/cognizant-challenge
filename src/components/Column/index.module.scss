.column{
    padding: 1rem;
    width: 100%;
   background-color: #eeeeee;
    &:not(:last-of-type){
      margin-right: 1rem;
    }
    & > h2{
      font-size: 2rem;
      margin-bottom: 3rem;
    }
    & > button{
     width: 100%;
    }
    & > p{
      font-size: 1.5rem;
      margin-bottom: 1.5rem;
      color: #666666;
    }
}