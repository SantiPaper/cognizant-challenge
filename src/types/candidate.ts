export interface Candidate {
  id: string;
  name: string;
  comments: string;
  step: Step;
}

export type Step =
  | "Entrevista inicial"
  | "Entrevista técnica"
  | "Oferta"
  | "Asignación"
  | "Rechazo";
