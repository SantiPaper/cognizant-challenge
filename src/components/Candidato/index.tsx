import React, {FunctionComponent} from "react";

import {Candidate} from "../../types/candidate";

import style from "./index.module.scss";

type Props = {
  candidato: Candidate;
  advance: (id: string, action: "avanzar" | "retroceder") => void;
};

export const Candidato: FunctionComponent<Props> = ({candidato, advance}) => {
  return (
    <article className={style.containerCandidato}>
      <div>
        <h3>{candidato.name}</h3>
        <p>{candidato.comments}</p>
      </div>
      <div>
        {candidato.step !== "Entrevista inicial" && (
          <button
            aria-label="Retroceder proceso"
            onClick={() => advance(candidato.id, "retroceder")}
          >
            {"<"}
          </button>
        )}
        {candidato.step !== "Rechazo" && (
          <button aria-label="Avanzar proceso" onClick={() => advance(candidato.id, "avanzar")}>
            {">"}
          </button>
        )}
      </div>
    </article>
  );
};
