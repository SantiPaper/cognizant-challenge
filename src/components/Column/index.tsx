import type {Candidate, Step} from "../../types/candidate";

import React, {FormEvent, FunctionComponent, useState} from "react";
import {v4} from "uuid";

import {Candidato} from "../Candidato";

import style from "./index.module.scss";

type Props = {
  candidatos: Candidate[];
  titulo: Step;
  showButton?: boolean;
  advance: (id: string, action: "avanzar" | "retroceder") => void;
  addCandidate: (candidato: Candidate) => void;
};

export const Column: FunctionComponent<Props> = ({
  candidatos,
  titulo,
  advance,
  showButton,
  addCandidate,
}) => {
  const [showForm, setShowForm] = useState(false);

  const crearCandidato = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const formData = new FormData(e.target);
    const newCandidate = Object.fromEntries(formData.entries());

    newCandidate.id = v4();
    newCandidate.step = "Entrevista inicial";

    addCandidate(newCandidate as unknown as Candidate);
  };

  return (
    <div className={style.column}>
      <h2>{titulo}</h2>

      {candidatos.length > 0 ? (
        candidatos.map((candidato) => (
          <Candidato key={candidato.id} advance={advance} candidato={candidato} />
        ))
      ) : (
        <p>No hay candidatos</p>
      )}

      {showButton && (
        <button
          onClick={() => {
            setShowForm(!showForm);
          }}
        >
          Agregar candidato
        </button>
      )}
      {showForm && (
        <form onSubmit={crearCandidato}>
          <label htmlFor="nombre">Nombre</label>
          <input id="nombre" name="name" type="text" />

          <label htmlFor="comentario">Algun comentario</label>
          <input id="comentario" name="comments" type="text" />

          <button>Dale loco</button>
        </form>
      )}
    </div>
  );
};
