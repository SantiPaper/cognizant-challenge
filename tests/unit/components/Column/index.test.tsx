import React from "react";
import {render, screen, act} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import {Candidate} from "../../../../src/types/candidate";
import {Column} from "../../../../src/components/Column";

vi.mock("uuid", () => {
  return {
    v4: () => "abc123",
  };
});

const candidates: Candidate[] = [
  {
    name: "Santiago",
    comments: "Burro",
    id: "Santi",
    step: "Entrevista inicial",
  },
  {
    name: "Emiliano",
    comments: "Crack",
    id: "Emi",
    step: "Entrevista inicial",
  },
];

const advance = () => {};

const addCandidate = vi.fn();

describe("Componente column", () => {
  const user = userEvent.setup();

  test("Se rendenderizan todos los participantes", () => {
    const {container} = render(
      <Column
        addCandidate={addCandidate}
        advance={advance}
        candidatos={candidates}
        titulo="Entrevista inicial"
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test("Si no hay participante renderiza texto alternativo", () => {
    render(
      <Column
        addCandidate={addCandidate}
        advance={advance}
        candidatos={[]}
        titulo="Entrevista técnica"
      />,
    );

    const noMatchCandidates = screen.getByText("No hay candidatos");

    expect(noMatchCandidates).toBeInTheDocument();
  });

  /*  test("Llega bien el titulo", () => {
    render(
      <Column
        addCandidate={addCandidate}
        advance={advance}
        candidatos={[]}
        titulo="Entrevista técnica"
      />,
    );

    const title = screen.getByRole("heading", {name: "Entrevista técnica"}); 

    expect(title).toBeVisible();
  });
  */

  test("Renderiza el boton del form solo en entrevista inicial", () => {
    render(
      <Column
        addCandidate={addCandidate}
        advance={advance}
        candidatos={candidates}
        titulo="Entrevista técnica"
      />,
    );
    const formButton = screen.queryByRole("button", {name: "Agregar candidato"});

    expect(formButton).not.toBeInTheDocument();
  });

  test("El boton agregar candidato funciona", () => {
    render(
      <Column
        addCandidate={addCandidate}
        advance={advance}
        candidatos={candidates}
        titulo="Entrevista inicial"
      />,
    );
    const addButton = screen.getByRole("button", {name: "Agregar candidato"});

    act(() => addButton.click());

    const form = screen.getByRole("form");

    expect(form).toBeVisible();

    act(() => addButton.click());
    expect(form).not.toBeInTheDocument();
  });

  test("Funciona el formulario", async () => {
    render(
      <Column
        addCandidate={addCandidate}
        advance={advance}
        candidatos={candidates}
        titulo="Entrevista inicial"
      />,
    );
    const showFormButton = screen.getByRole("button", {name: "Agregar candidato"});

    act(() => showFormButton.click());

    const inputName = screen.getByRole("textbox", {name: "Nombre"});
    const inputComment = screen.getByRole("textbox", {name: "Algun comentario"});
    const sendFormButton = screen.getByRole("button", {name: "Agregar"});

    sendFormButton.click();

    expect(addCandidate).not.toHaveBeenCalled();

    await user.type(inputName, "Gonzalo");
    await user.type(inputComment, "Medio pelo");

    sendFormButton.click();

    expect(addCandidate).toHaveBeenCalledWith({
      name: "Gonzalo",
      comments: "Medio pelo",
      step: "Entrevista inicial",
      id: "abc123",
    });
  });
});
