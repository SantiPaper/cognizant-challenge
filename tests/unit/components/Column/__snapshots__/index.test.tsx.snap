// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Componente column > Se rendenderizan todos los participantes 1`] = `
<div>
  <div
    class="_column_545bf0"
  >
    <h2>
      Entrevista inicial
    </h2>
    <article
      class="_containerCandidato_59e197"
    >
      <div>
        <h3>
          Santiago
        </h3>
        <p>
          Burro
        </p>
      </div>
      <div>
        <button
          aria-label="Avanzar proceso"
        >
          &gt;
        </button>
      </div>
    </article>
    <article
      class="_containerCandidato_59e197"
    >
      <div>
        <h3>
          Emiliano
        </h3>
        <p>
          Crack
        </p>
      </div>
      <div>
        <button
          aria-label="Avanzar proceso"
        >
          &gt;
        </button>
      </div>
    </article>
    <button>
      Agregar candidato
    </button>
  </div>
</div>
`;
